File: BaRia_2324/Bai3/REHALL.py
input_path = './THCS/2024/BaRia/Bai3/REHALL.INP'
output_path = './THCS/2024/BaRia/Bai3/REHALL.OUT'
def main():
    with open(input_path, 'r') as file:
        n = int(file.readline().strip())
        events = []
        for _ in range(n):
            start, end, cost = map(int, file.readline().strip().split())
            events.append((start, end, cost))

    # Sắp xếp các sự kiện theo thời gian kết thúc
    events.sort(key=lambda x: x[1])

    # dp lưu lợi nhuận tối ưu tại thời điểm kết thúc
    dp = [0] * n  # dp[i] là lợi nhuận tối ưu tại sự kiện thứ i
    prev = [-1] * n  # prev[i] là sự kiện không trùng thời gian với sự kiện i

    # Tìm sự kiện trước đó không trùng với sự kiện hiện tại
    for i in range(n):
        for j in range(i - 1, -1, -1):
            if events[j][1] < events[i][0]:  # sự kiện j kết thúc trước khi sự kiện i bắt đầu
                prev[i] = j
                break

    # Tính toán lợi nhuận tối ưu
    for i in range(n):
        if prev[i] == -1:
            dp[i] = events[i][2]  # Nếu không có sự kiện nào trước đó, lợi nhuận chỉ là chính sự kiện đó
        else:
            dp[i] = max(dp[i - 1], dp[prev[i]] + events[i][2])  # Chọn giữa việc lấy hoặc không lấy sự kiện i

    max_profit = max(dp)

    with open(output_path, 'w') as file:
        file.write(f"{max_profit}\n")

if __name__ == "__main__":
    main()

File: THCS/2023_2024/BaRia_2324/Bai3/REHALL.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

struct Event {
    int start, end, cost;
};

// So sánh các sự kiện theo thời gian kết thúc
bool compare(Event &a, Event &b) {
    return a.end < b.end;
}

int main() {
    ifstream input("REHALL.INP");
    ofstream output("REHALL.OUT");

    int n;
    input >> n;

    vector<Event> events(n);
    for (int i = 0; i < n; ++i) {
        input >> events[i].start >> events[i].end >> events[i].cost;
    }

    // Sắp xếp các sự kiện theo thời gian kết thúc
    sort(events.begin(), events.end(), compare);

    // Sử dụng map để lưu thời gian kết thúc và tổng chi phí tối ưu
    map<int, long long> dp;
    dp[0] = 0; // Khởi tạo trạng thái ban đầu

    long long max_profit = 0;

    for (auto &e : events) {
        // Tìm sự kiện có thể chọn trước sự kiện hiện tại mà không trùng thời gian
        auto it = dp.lower_bound(e.start);
        --it; // Lùi lại để lấy giá trị nhỏ hơn hoặc bằng e.start
        long long current_profit = it->second + e.cost;
        
        // Cập nhật tổng lợi nhuận tối ưu tại thời điểm e.end
        if (current_profit > dp[e.end]) {
            dp[e.end] = current_profit;
            max_profit = max(max_profit, current_profit);
        }
    }

    // Ghi kết quả ra file
    output << max_profit << endl;

    input.close();
    output.close();

    return 0;
}
